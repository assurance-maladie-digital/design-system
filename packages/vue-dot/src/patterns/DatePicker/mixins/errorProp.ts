import Vue from 'vue';
import Component, { mixins } from 'vue-class-component';

const Props = Vue.extend({
	props: {
		error: {
			type: Boolean,
			default: false
		}
	}
});

const MixinsDeclaration = mixins(Props);

@Component
export class ErrorProp extends MixinsDeclaration {
	get internalErrorProp(): boolean {
		return this.error;
	}

	set internalErrorProp(value: boolean) {
		this.$emit('update:error', value);
	}
}
