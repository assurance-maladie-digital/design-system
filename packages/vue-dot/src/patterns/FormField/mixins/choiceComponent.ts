import Vue, { PropType } from 'vue';
import Component from 'vue-class-component';

import { ChoiceValue, FieldItem, FieldMetadata, FieldItemValue } from '../types';

const Props = Vue.extend({
	props: {
		/** The choice field to display */
		value: {
			type: [Array, Number, String] as PropType<ChoiceValue>,
			default: null
		},
		metadata: {
			type: Object as PropType<FieldMetadata>,
			default: null
		},
		multiple: {
			type: Boolean,
			default: false
		},
		items: {
			type: Array as PropType<FieldItem[]>,
			required: true
		}
	}
});

/** Mixin to control the item selection */
@Component<ChoiceComponent>({
	watch: {
		// Listen the current field value for the component
		value: {
			handler(value: ChoiceValue) {
				if (value !== null && value !== undefined) {
					/** In multiple mode, put the value in an array if it wasn't already */
					if (this.multiple && !Array.isArray(value)) {
						this.choiceValue = [value];
					} else {
						this.choiceValue = value;
					}
				} else {
					// Reset the choice value
					this.choiceValue = this.multiple ? [] : null;
				}
			},
			immediate: true,
			deep: true
		}
	}
})
export class ChoiceComponent extends Props {
	choiceValue: ChoiceValue | null = this.multiple ? [] : null;

	/**
	 * Toggle the item
	 *
	 * @param {FieldItem} item The selected item
	 * @returns {void}
	 */
	toggleItem(item: FieldItem): void {
		const active: Boolean = this.isSelected(item.value);

		let newChoiceValue: ChoiceValue;

		// Set the new value in simple mode
		if (!this.multiple) {
			newChoiceValue = active ? null : item.value;
		} else {
			newChoiceValue = [
				...this.choiceValue as FieldItemValue[]
			];

			if (active && Array.isArray(this.choiceValue)) {
				// Unselect the item
				const valueIndex = this.choiceValue.indexOf(item.value);

				newChoiceValue.splice(valueIndex, 1);
			} else {
				// Can't select a null value in multiple mode
				if (item.value === undefined || item.value === null) {
					return;
				}

				const isAlone: boolean = Boolean(item.alone);

				// If item alone, unselect all other
				if (isAlone) {
					newChoiceValue = [item.value];
				} else {
					// Else unselect all alone items
					let index = newChoiceValue.length - 1;

					while (index >= 0) {
						const valueSelected = (newChoiceValue as FieldItemValue[])[index];

						// Search if the item is alone
						const isItemSelectedAlone = this.items.find((element) => {
							return element.value === valueSelected && element.alone;
						});

						if (isItemSelectedAlone) {
							// Delete alone item
							newChoiceValue.splice(index, 1);
						}

						index -= 1;
					}

					// Add the item to the selection
					newChoiceValue.push(item.value);
				}
			}
		}

		this.choiceValue = newChoiceValue;

		this.emitChoiceUpdated(this.choiceValue);
	}

	emitChoiceUpdated(newValue: ChoiceValue) {
		this.$emit('change', newValue);
	}

	/**
	 * Check if the button is selected
	 *
	 * @param {FieldItemValue} value The button value to test if it is selected
	 * @returns {boolean} The selected value
	 */
	isSelected(value: FieldItemValue): boolean {
		if (!this.choiceValue) {
			return false;
		}

		if (Array.isArray(this.choiceValue)) {
			return this.choiceValue.includes(value);
		} else {
			return this.choiceValue === value;
		}
	}
}
