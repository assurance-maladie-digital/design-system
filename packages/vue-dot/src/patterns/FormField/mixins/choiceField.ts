import Component from 'vue-class-component';

import { ChoiceValue, FieldItem } from '../types';

import FieldComponent from './fieldComponent';

/** Choice field component to controle the item selection if needed */
@Component<ChoiceField>({
	watch: {
		// Listen the current field value for the component
		'field.value': {
			handler(value: ChoiceValue) {
				if (value) {
					this.choiceValue = value;
				}
			},
			immediate: true,
			deep: true
		}
	}
})
export default class ChoiceField extends FieldComponent {
	choiceValue?: ChoiceValue;

	get isMultiple() {
		return this.field && this.field.metadata && this.field.metadata.multiple;
	}

	/**
	 * Toggle the item
	 *
	 * @param {FieldItem} item The selected item
	 * @param {boolean} active Is selected or not
	 */
	toggleItem(item: FieldItem, active: boolean) {
		/**
		 * Items must be an array
		 * Item value must be defined
		 */
		if (!Array.isArray(this.field.items)) {
			return;
		}
		// Set the new value in simple mode
		if (!this.isMultiple) {
			this.choiceValue = active ? null : item.value;
		} else if (active && Array.isArray(this.choiceValue)) {
			// Unselect the item
			const valueIndex = this.choiceValue.indexOf(item.value);
			this.choiceValue.splice(valueIndex,1);
		} else {

			// Can't select a null value in multiple mode
			if (!item.value) {
				return;
			}
			// Init choiceValue as array if not
			if (!Array.isArray(this.choiceValue)) {
				this.choiceValue = [];
			}

			const isAlone: boolean = Boolean(item.alone);

			// If item alone, unselect all other
			if (isAlone) {
				this.choiceValue = [item.value];
			} else {
				// Else unselect all alone
				let index = this.choiceValue.length - 1;
				while (index >= 0) {
					const valueSelected = this.choiceValue[index];
					// Search if the item is alone
					const isItemSelectedAlone = this.field.items.find((element) => {
						return element.value === valueSelected && element.alone;
					});
					if (
						isItemSelectedAlone
					) {
						// Delete alone item
						this.choiceValue.splice(index, 1);
					}

					index -= 1;
				}
				// Add the item to the selection
				this.choiceValue.push(item.value);
			}
		}

		this.emitChangeEvent(this.choiceValue);
	}
}
