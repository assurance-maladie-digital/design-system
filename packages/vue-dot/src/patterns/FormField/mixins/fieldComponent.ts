import Vue from 'vue';
import Component from 'vue-class-component';

import { Field } from '../types';

const Props = Vue.extend({
	props: {
		/** The field to display */
		field: {
			type: [Array, Object],
			required: true
		}
	}
});

/** Share code between field components */
@Component({
	model: {
		prop: 'field',
		event: 'change'
	}
})
export default class FieldComponent extends Props {
	// Stronger types
	field!: Field;

	/**
	 * Update the v-model by emitting 'change' event
	 *
	 * @param {Field} field The updated field
	 * @returns {void}
	 */
	emitChangeEvent(field: Field): void {
		// Emit in next tick to respect event order
		this.$nextTick(() => {
			this.$emit('change', field);
		});
	}
}
