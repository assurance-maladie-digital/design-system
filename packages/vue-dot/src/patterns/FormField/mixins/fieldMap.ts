import Vue from 'vue';
import Component from 'vue-class-component';

interface IFieldMap {
	[key: string]: string;
}

// We import them all because the form
// can use any of them
import AutocompleteField from '../fields/AutocompleteField.vue';
import DateField from '../fields/DateField.vue';
import FileUploadField from '../fields/FileUploadField';
import NumberField from '../fields/NumberField.vue';
import PasswordField from '../fields/PasswordField.vue';
import PeriodField from '../fields/PeriodField.vue';
import SelectField from '../fields/SelectField.vue';
import ChoiceButtonField from '../fields/ChoiceButtonField.vue';
import ChoiceSliderField from '../fields/ChoiceSliderField.vue';
import TextareaField from '../fields/TextareaField.vue';
import TextField from '../fields/TextField.vue';

/** List all fields and provide getField() function */
@Component({
	components: {
		AutocompleteField,
		DateField,
		FileUploadField,
		NumberField,
		PasswordField,
		PeriodField,
		SelectField,
		ChoiceButtonField,
		ChoiceSliderField,
		TextareaField,
		TextField
	}
})
export class FieldMap extends Vue {
	/** List all field components and their corresponding keys */
	fieldMap: IFieldMap = {
		autocomplete: 'AutocompleteField',
		date: 'DateField',
		fileUpload: 'FileUploadField',
		number: 'NumberField',
		password: 'PasswordField',
		period: 'PeriodField',
		select: 'SelectField',
		choiceButton: 'ChoiceButtonField',
		choiceSlider: 'ChoiceSliderField',
		text: 'TextField',
		textarea: 'TextareaField'
	};

	/**
	 * Returns the field that correspond to the name
	 * passed in argument
	 *
	 * @param {string} fieldName The name of the field
	 * @returns {string} The field
	 */
	getField(fieldName: string): string {
		return this.fieldMap[fieldName];
	}
}
