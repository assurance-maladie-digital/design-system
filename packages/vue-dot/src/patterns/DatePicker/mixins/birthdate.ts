import dayjs from 'dayjs';
import Vue from 'vue';
import Component, { mixins } from 'vue-class-component';

import { INTERNAL_FORMAT } from './dateLogic';

const Props = Vue.extend({
	props: {
		birthdate: {
			type: Boolean,
			default: false
		}
	}
});

const MixinsDeclaration = mixins(Props);

@Component<Birthdate>({
	watch: {
		menu(value: boolean): void {
			if (this.birthdate && value) {
				this.setActivePicker();
			}
		}
	}
})
export class Birthdate extends MixinsDeclaration {
	activePicker: string | null = null;

	max = this.birthdate ? dayjs().format(INTERNAL_FORMAT) : null;

	min = this.birthdate ? '1900-01-01' : null;

	/** @see https://vuetifyjs.com/en/components/date-pickers#date-pickers-birthday-picker */
	setActivePicker(): void {
		setTimeout(() => {
			this.activePicker = 'YEAR';
		});
	}
}
