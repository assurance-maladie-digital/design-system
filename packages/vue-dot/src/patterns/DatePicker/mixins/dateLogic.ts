import Vue from 'vue';
import Component, { mixins } from 'vue-class-component';

import { parseDate } from '../../../helpers/parseDate';
import { DATE_FORMAT_REGEX } from '../../../functions/validation/isDateValid';

import { Options } from '../../../mixins/customizable';

import { Refs } from '../../../types';

/** Date format used internally */
const INTERNAL_FORMAT = 'YYYY-MM-DD';

const Props = Vue.extend({
	props: {
		/** The format of the date inside the text field */
		dateFormat: {
			type: String,
			default: 'DD/MM/YYYY'
		},
		/** The format used in the v-model for the user */
		dateFormatReturn: {
			type: String,
			default: 'YYYY-MM-DD'
		},
		/** The v-model value */
		value: {
			type: String,
			default: ''
		}
	}
});

const MixinsDeclaration = mixins(Props);

/** Handle main logic of the DatePicker */
@Component<DateLogic>({
	watch: {
		/** Update the date when value is provided by the user */
		value: {
			handler(date: string): void {
				// If the date is cleared
				if (!date) {
					// Clear internal models
					this.date = '';
					this.textFieldDate = '';

					return;
				}

				// Format the date to internal format using dateFormatReturn
				const parsed = this.parseDateForModel(date);

				// If parsed is an empty string, the date isn't valid, don't continue
				if (!parsed) {
					return;
				}

				this.date = parsed;
				this.setTextFieldModel();

				// Validate warning rules
				this.validate(this.textFieldDate);
				// Validate Vuetify rules
				this.validateVuetify();
			},
			immediate: true
		},
		/**
		 * This method is fired every time textFieldDate changes,
		 * that means on every key strokes, so don't validate
		 * on input if validateOnBlur is true
		 */
		textFieldDate(value: string): void {
			if (!this.validateOnBlurEnabled) {
				this.validate(value);
			}
		}
	}
})
export class DateLogic extends MixinsDeclaration {
	// Extend $refs
	$refs!: Refs<{
		/** VMenu */
		menu: {
			save: (date: string) => void;
		};
		/** VTextField */
		input: {
			validate: () => boolean;
			hasFocused: boolean;
			hasError: boolean;
		};
	}>;

	// DatePicker.options
	options!: Options;

	// Mixin computed data
	// warning rules
	validate!: (value: string) => void;

	/**
	 * The v-model of the component
	 *
	 * @example
	 * Format is '2018-03-25'
	 *
	 * Set initial value by parsing if there is one
	 */
	date = this.value ? this.parseDateForModel(this.value) : '';

	/**
	 * The v-model of the text field
	 * it's different od this.date because the formatting
	 * isn't the same
	 *
	 * @example
	 * Format is '25032018'
	 */
	textFieldDate = '';

	mounted() {
		// Watch VTextField 'hasError' computed value
		// because 'update:error' event isn't reliable
		// (it's not fired at initial state and 'validateOnBlur'
		// can cause issues as well)
		this.$watch(
			() => this.$refs.input.hasError,
			(error: boolean) => {
				this.$emit('error', error);
			},
			{
				deep: true // Even if we don't watch an object, this is needed
			}
		);
	}

	/** Check if validateOnBlur is enabled */
	get validateOnBlurEnabled(): boolean {
		return Boolean(this.options.textField?.validateOnBlur);
	}

	/** Parse a date with dateFormatReturn format to internal format */
	parseDateForModel(date: string): string {
		const parsed = parseDate(date, this.dateFormatReturn);

		if (!parsed.isValid()) {
			return '';
		}

		return parsed.format(INTERNAL_FORMAT);
	}

	parseTextFieldDate(date: string): string {
		const formatted = parseDate(date, this.dateFormat);

		// If the date isn't valid, return empty string
		if (!date.match(DATE_FORMAT_REGEX) || !formatted.isValid()) {
			return '';
		}

		return formatted.format(INTERNAL_FORMAT);
	}

	/** Set textField model by parsing this.date */
	setTextFieldModel(): void {
		this.textFieldDate = parseDate(this.date, INTERNAL_FORMAT).format(this.dateFormat);
	}

	/**
	 * Format date with dayjs and dateFormat
	 *
	 * @example
	 * Format is '25/03/2018' with default dateFormat
	 */
	// Getter
	get dateFormatted(): string {
		/**
		 * If the date is empty, return now
		 * to avoid date parsing errors
		 */
		if (this.date === '') {
			return '';
		}
		/** Format this.date with dateFormat */
		const formatted = parseDate(this.date, 'YYYY-MM-DD').format(this.dateFormat);


		return formatted;
	}

	// Setter
	set dateFormatted(value: string) {
		this.textFieldDate = value;
	}

	/** Save the date from calendar */
	saveFromCalendar(): void {
		// Save the date using Vuetify method
		this.$refs.menu.save(this.date);

		// Update textField model
		this.setTextFieldModel();

		// Apply validation because when the calendar is clicked,
		// the input loose focus and fire textFieldBlur
		this.validate(this.textFieldDate);

		// If validateOnBlur is true
		if (this.validateOnBlurEnabled) {
			// Validate the text field
			// because no blur event is emitted
			this.validateVuetify();
		}

		this.emitChangeEvent();
	}

	/** Save the date from text field blur */
	saveFromTextField(): void {
		// Return if empty/falsy
		if (!this.textFieldDate) {
			// Clear v-model
			this.$emit('change', '');
			return;
		}

		const formatted = this.parseTextFieldDate(this.textFieldDate);

		// Set the internal date
		this.date = formatted;

		this.emitChangeEvent();
	}

	/** Update v-model */
	emitChangeEvent(): void {
		// Parse the date with internal format,
		// and return the date with dateFormatReturn format
		this.$emit('change', parseDate(this.date, INTERNAL_FORMAT).format(this.dateFormatReturn));
	}

	/** Validate Vuetify rules */
	validateVuetify(): void {
		this.$nextTick(() => {
			// Set hasFocused to true on VTextField
			// to fix https://github.com/vuetifyjs/vuetify/issues/7876
			// (error messages aren't shown if the input hasn't been focused)
			if (this.$refs.input.hasFocused !== undefined) {
				this.$refs.input.hasFocused = true;
			}

			this.$refs.input.validate();
		});
	}

	/** Fired on blur event of the textField */
	textFieldBlur(): void {
		this.saveFromTextField();

		// If validateOnBlur is true, validate
		if (this.validateOnBlurEnabled) {
			this.validate(this.textFieldDate);
		}
	}
}
