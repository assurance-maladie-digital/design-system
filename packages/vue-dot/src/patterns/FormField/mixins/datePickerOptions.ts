import Vue from 'vue';
import Component from 'vue-class-component';

import { Options } from '../../../mixins/customizable';

import merge from 'deepmerge';

/** Merge default options (eg. menu position fix) with metadata */
@Component
export class DatePickerOptions extends Vue {
	/**
	 * Get Vuetify options object
	 *
	 * @param {Options} metadata The metadata to merge with
	 * @returns {Options} The merged options
	 */
	getVuetifyOptions(metadata?: Options): Options {
		const defaultOptions = {
			menu: {
				nudgeBottom: 55,
				nudgeRight: 0
			}
		};

		if (metadata) {
			return merge(defaultOptions, metadata) as unknown as Options;
		}

		return defaultOptions;
	}
}
