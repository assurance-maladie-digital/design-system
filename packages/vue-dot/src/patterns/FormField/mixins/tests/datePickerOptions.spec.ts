import Vue from 'vue';
import { mount, Wrapper } from '@vue/test-utils';

import { Options } from '../../../../mixins/customizable';

import { DatePickerOptions } from '../datePickerOptions';

/** Create the wrapper */
function createWrapper() {
	const component = Vue.component('test', {
		mixins: [
			DatePickerOptions
		],
		template: '<div />'
	});

	return mount(component) as Wrapper<DatePickerOptions>;
}

// Tests
describe('DatePickerOptions', () => {
	it('returns the options', () => {
		const wrapper = createWrapper();

		expect(wrapper.vm.getVuetifyOptions()).toMatchSnapshot();
	});

	it('returns the options and merge correctly', () => {
		const wrapper = createWrapper();

		const options: Options = {
			menu: {
				nudgeBottom: 80
			},
			other: {
				test: true
			}
		};

		expect(wrapper.vm.getVuetifyOptions(options)).toMatchSnapshot();
	});
});
