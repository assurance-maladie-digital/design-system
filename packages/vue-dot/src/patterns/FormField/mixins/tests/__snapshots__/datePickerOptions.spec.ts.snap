// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datePickerOptions returns the options 1`] = `
Object {
  "menu": Object {
    "nudgeBottom": 55,
    "nudgeRight": 0,
  },
}
`;

exports[`datePickerOptions returns the options and merge correctly 1`] = `
Object {
  "menu": Object {
    "nudgeBottom": 80,
    "nudgeRight": 0,
  },
  "other": Object {
    "test": true,
  },
}
`;
